### PR TITLE
IFU to v2.0.4

### PR DESCRIPTION
**Pull request by howiejayz**
- [x] renamed "unpadded" -> "varlen" with mha_varlen_fwd & mha_varlen_bwd APIs.
- [x] changed mha_fwd & mha_bwd for input with the same sequence lengths in the same batch.
- [x] setup.py will install for either CUDA/ROCm system.
- [x] rename test_flash_attn -> test_flash_attn_rocm for ROCm unit test.
- [x] benchmark testing
- [x] Sync to PR #15 
- [x] add unit tests for mha_fwd&mha_bwd
- [x] MQA/GQA
- [ ] unit test all pass

**Current Unit Test Result:** (PyTorch 2.0.0; ROCm 5.6)
3968 passed, 63 skipped

**Current Performance on MI250**: (docker pull rocm/pytorch:rocm5.7_ubuntu22.04_py3.10_pytorch_2.0.1)
  | fwd tflops | bwd tflops | total
-- | -- | -- | --
fp16 | 52.16 | 39.93 | 42.49
bf16 | 52.36 | 30.25 | 34.21

### CODE REVIEW

**Comment by fsx950223** [comment]
So big.

File: csrc/flash_attn_rocm/flash_api.cpp [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+                               p_dropout,
+                               softmax_scale,
+                               is_causal,
+                               return_softmax);
+
+  // number of times random will be generated per thread, to offset philox counter in thc random
+  // state
+  // We use a custom RNG that increases the offset by batch_size * nheads * 32.
+  int64_t counter_offset = params.b * params.h_q * 32;
+  auto options = torch::TensorOptions().dtype(torch::kFloat32).device(torch::kCUDA);
```
Comment (fsx950223) [inline csrc/flash_attn_rocm/flash_api.cpp:278]: Two dtypes, use one instead.

File: csrc/flash_attn_rocm/flash_api.cpp [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
```
Comment (fsx950223) [inline csrc/flash_attn_rocm/flash_api.cpp:120]: Two dtypes, use one instead.

**Comment by regnnighe** [comment]
A new environment variable "FLASH_ATTENTION_INTERNAL_ENABLE_TIME_KERNEL" can switch the output of kernel running time

**Comment by regnnighe** [comment]
[BUGs] Previously in older version of FA, we create tensors z and softmax_lse matrix of max sseqence lengths with no padding for grouped gemm. But the strides for each batch for the tensors are different. This behaviour will cause wrong result from CK. Fixing this.

**Comment by fsx950223** [comment]
Please remove *_hip.hpp

File: .gitignore [hunk 3]
```diff
@@ -24,7 +24,10 @@ var/
 .vscode/settings.
 
 # Generated files
+csrc/flash_attn_rocm/src/*hip*
```
Comment (fsx950223) [inline .gitignore:27]: beter to use *_hip*?

Comment (regnnighe) [reply .gitignore:27]: makes sense

Comment (fsx950223) [reply .gitignore]: Yes, file names such as ```hip_flash_attention.cpp``` are ignored too?
